url1f1de5fc71_cooijmanstim_hobo_tgz-pJ8-hobo_TestingEnviromentJ8/src/hobo/Belief.java

```diff
@@ -0,0 +1,636 @@
+			double[][] jpd = Util.clone(player_mission_suspicion);
+			for (int i = 0; i < jpd.length; i++) {
+				for (int j = 0; j < jpd[i].length; j++) {
+					// more probability to the players that need more missions,
+					// less probability to those that have more railways (to avoid bias,
+					// and to generally take the average of the relevance of the claimed
+					// railways)
+					if (ns[j] == 0 && jpd[i][j] == Double.POSITIVE_INFINITY) {
+						System.err.println("inconsistent certainty");
+						throw new RuntimeException();
+					}
+					jpd[i][j] *= ns[j];
+					if (jpd[i][j] < 0 || Double.isNaN(jpd[i][j])) {
+						System.err.println("negative or NaN weight in non-normalized distribution: "+jpd[i][j]+" (originally "+player_mission_suspicion[i][j]+")");
+						System.err.println("last factor: "+ns[j]+"/"+(1 + s.playerState(j).railways.size()));
+						System.err.println("original distribution: "+Arrays.deepToString(player_mission_suspicion));
+						throw new RuntimeException();
+					}
+					total += jpd[i][j];
+				}
+			}
+			if (total < 0 || Double.isNaN(total)) {
+				System.err.println("non-normalized probability distribution sums to "+total);
+				System.err.println("original distribution: "+Arrays.deepToString(player_mission_suspicion));
+				System.err.println("jpd: "+Arrays.deepToString(jpd));
+				throw new RuntimeException();
+			}
+
+			// now draw a sample from the jpd
+			double x = random.nextDouble() * total;
+			for (int i = 0; i < jpd.length; i++) {
+				for (int j = 0; j < jpd[i].length; j++) {
+					x -= jpd[i][j];
+					// positive infinity signifies certainty but will result in
+					// x being NaN, so special-case it
+					if (x <= 0 || jpd[i][j] == Double.POSITIVE_INFINITY) {
+						Mission m = Mission.all[i];
+						s.missions.remove(m);
+						s.playerState(j).missions.add(m);
+						ns[j]--; // one fewer mission needed for this player
+						// this mission will no longer be sampled
+						for (int k = 0; k < players.length; k++)
+							player_mission_suspicion[i][k] = 0;
+						continue sampling;
+					}
+					if (Double.isNaN(x)) {
+						System.out.println("x just became NaN, last subtraction: "+jpd[i][j]);
+						System.out.println("original distribution: "+Arrays.deepToString(player_mission_suspicion));
+						System.out.println("jpd: "+Arrays.deepToString(jpd));
+						throw new RuntimeException();
+					}
+				}
+			}
+
+			// shouldn't get here
+			System.err.println(x+" of mass left. distribution:");
+			System.err.println(Arrays.deepToString(jpd));
+			throw new RuntimeException();
+		}
+
+		// and this took me DAYS to find -- make sure drawn_missions is consistent as well
+		for (PlayerState ps: s.playerStates()) {
+			// drawn_missions for ourselves are done above, before sampling
+			if (ps.handle == player)
+				continue;
+			if (ps.drawn_missions != null)
+				ps.drawn_missions = Util.remove_sample(s.missions, ps.drawn_missions.size(), random, EnumSet.noneOf(Mission.class));
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 411:417, 498:504.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.